File: 3312. sorted-gcd-pair-queries.java
import java.util.Arrays;

class Solution {
    public int[] gcdValues(int[] nums, long[] queries) {
        int maxVal = 0;
        for (int x : nums) maxVal = Math.max(maxVal, x);

        // frequency of each number
        long[] freq = new long[maxVal + 1];
        for (int x : nums) freq[x]++;

        // divCnt[g] = how many numbers are divisible by g
        long[] divCnt = new long[maxVal + 1];

        for (int g = 1; g <= maxVal; g++) {
            for (int x = g; x <= maxVal; x += g) {
                divCnt[g] += freq[x];
            }
        }

        // exact[g] = number of pairs whose gcd is exactly g
        long[] exact = new long[maxVal + 1];

        for (int g = maxVal; g >= 1; g--) {

            // choose any two numbers divisible by g
            exact[g] = divCnt[g] * (divCnt[g] - 1) / 2;

            // remove pairs whose gcd is a larger multiple of g
            for (int m = 2 * g; m <= maxVal; m += g) {
                exact[g] -= exact[m];
            }
        }

        // prefix[g] = number of pairs with gcd <= g
        long[] prefix = new long[maxVal + 1];

        for (int g = 1; g <= maxVal; g++) {
            prefix[g] = prefix[g - 1] + exact[g];
        }

        int[] ans = new int[queries.length];

        for (int i = 0; i < queries.length; i++) {
            long target = queries[i] + 1;

            int left = 1, right = maxVal;

            while (left < right) {
                int mid = left + (right - left) / 2;

                if (prefix[mid] >= target)
                    right = mid;
                else
                    left = mid + 1;
            }

            ans[i] = left;
        }

        return ans;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        // Example 1
        int[] nums1 = {2, 3, 4};
        long[] queries1 = {0, 2, 2};
        System.out.println(Arrays.toString(sol.gcdValues(nums1, queries1)));
        // [1, 2, 2]

        // Example 2
        int[] nums2 = {4, 4, 2, 1};
        long[] queries2 = {5, 3, 1, 0};
        System.out.println(Arrays.toString(sol.gcdValues(nums2, queries2)));
        // [4, 2, 1, 1]

        // Example 3
        int[] nums3 = {2, 2};
        long[] queries3 = {0, 0};
        System.out.println(Arrays.toString(sol.gcdValues(nums3, queries3)));
        // [2, 2]
    }
}

File: 3312. sorted-gcd-pair-queries.py
from bisect import bisect_left
from typing import List


class Solution:
    def gcdValues(self, nums: List[int], queries: List[int]) -> List[int]:
        maxVal = max(nums)

        # frequency of each number
        freq = [0] * (maxVal + 1)
        for x in nums:
            freq[x] += 1

        # divCnt[g] = how many numbers are divisible by g
        divCnt = [0] * (maxVal + 1)

        for g in range(1, maxVal + 1):
            for x in range(g, maxVal + 1, g):
                divCnt[g] += freq[x]

        # exact[g] = number of pairs whose gcd is exactly g
        exact = [0] * (maxVal + 1)

        for g in range(maxVal, 0, -1):

            # choose any two numbers divisible by g
            exact[g] = divCnt[g] * (divCnt[g] - 1) // 2

            # remove pairs whose gcd is a larger multiple of g
            for m in range(2 * g, maxVal + 1, g):
                exact[g] -= exact[m]

        # prefix[g] = number of pairs with gcd <= g
        prefix = [0] * (maxVal + 1)

        for g in range(1, maxVal + 1):
            prefix[g] = prefix[g - 1] + exact[g]

        ans = []

        for q in queries:
            g = bisect_left(prefix, q + 1, 1)
            ans.append(g)

        return ans


if __name__ == "__main__":
    sol = Solution()

    # Example 1
    nums = [2, 3, 4]
    queries = [0, 2, 2]
    print(sol.gcdValues(nums, queries))  # [1, 2, 2]

    # Example 2
    nums = [4, 4, 2, 1]
    queries = [5, 3, 1, 0]
    print(sol.gcdValues(nums, queries))  # [4, 2, 1, 1]

    # Example 3
    nums = [2, 2]
    queries = [0, 0]
    print(sol.gcdValues(nums, queries))  # [2, 2]